src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+            if(parentHier == null)
+            {
+                ErrorHandler.ThrowOnFailure(VSConstants.E_UNEXPECTED);
+            }
+            uint parentItemid  = (uint)project.VsHierarchy.GetProperty<int>(VsHierarchyPropID.ParentHierarchyItemid, unchecked((int)VSConstants.VSITEMID_NIL));
+            if(parentItemid == VSConstants.VSITEMID_NIL)
+            {
+                ErrorHandler.ThrowOnFailure(VSConstants.E_UNEXPECTED);
+            }
+            
+            // Now using IVsPersistHierarchyItem2 we reload the project.
+            int hr = ((IVsPersistHierarchyItem2)parentHier).ReloadItem((uint)parentItemid, dwReserved: 0);
+            ErrorHandler.ThrowOnFailure(hr);
+        }
+
+        /// <summary>
+        /// Saves just the project file (does not save dirty documents)
+        /// </summary>
+       int SaveProject(IReloadableProject project)
+        {
+            IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+            __VSSLNSAVEOPTIONS saveOpts = __VSSLNSAVEOPTIONS.SLNSAVEOPT_SkipDocs;
+
+            IVsHierarchy hier;
+            uint itemid;
+            IVsPersistDocData docData;
+            uint docCookie;
+            VsShellUtilities.GetRDTDocumentInfo(_serviceProvider, project.ProjectFile, out hier, out itemid, out docData, out docCookie);
+            int hr = solution.SaveSolutionElement((uint)saveOpts, project.VsHierarchy, docCookie);
+            return hr;
+        }
+
+        /// <summary>
+        /// Uses SaveAs to save a copy of the project somewhere else.
+        /// </summary>
+        int SaveAsProject(IReloadableProject project)
+        {
+            // Save as needs to go through IPersistFileFormat
+            var persistFileFmt = project.VsHierarchy as IPersistFileFormat;
+            var uishell = _serviceProvider.GetService<IVsUIShell, SVsUIShell>();
+            string newFile;
+            int canceled;
+            int hr = uishell.SaveDocDataToFile(VSSAVEFLAGS.VSSAVE_SaveCopyAs, persistFileFmt, project.ProjectFile, out newFile, out canceled);
+
+            if (ErrorHandler.Succeeded(hr) && canceled == 1)
+            {
+                hr = VSConstants.E_ABORT;
+            }
+            return hr;
+        }
+
+        /// <summary>
+        /// Callback for directory changes. Since we don't watch the folder there is nothing to do here
+        /// </summary>
+        public int DirectoryChanged(string pszDirectory)
+        {
+            return VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// IVsSolutionEvents4. We only care about OnAfterRenameProject. If our project file is renamed we need to
+        /// stop watching the old file and start watching the new file
+        /// </summary>
+        public int OnAfterRenameProject(IVsHierarchy pHierarchy)
+        {
+            lock(_registeredProjects)
```

Comment:
We're not locking on `_registeredProjects` anywhere else, as far as I can tell. What does doing this here buy us?